Avoid confusion in Reflection with hidden members.

AaronRobinsonMSFT opened the pull request:
On CoreCLR, the Reflection API was preemptively
filtering out members with the same name and signature.
This means that hidden private members in C# using
the new keyword in derived classes can make public
members in base classes inaccessible.

Add tests for properties and methods.

Fixes https://github.com/dotnet/runtime/issues/28056
Fixes https://github.com/dotnet/runtime/issues/98533
Fixes https://github.com/dotnet/runtime/issues/118677

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/PropertyInfoTests.cs:
```diff
@@ -342,6 +342,14 @@ public void PropertyType(Type type, string name, Type expected)
+        [InlineData(typeof(SubClass), nameof(SubClass.BoolProperty))]
+        public void HiddenPublicProperty(Type type, string name)
+        {
+            Assert.NotNull(type.GetProperty(name));
```
Do we want to validate that it is property on the BaseClass?

jkotas commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/PropertyInfoTests.cs:
```diff
@@ -342,6 +342,14 @@ public void PropertyType(Type type, string name, Type expected)
+        [InlineData(typeof(SubClass), nameof(SubClass.BoolProperty))]
+        public void HiddenPublicProperty(Type type, string name)
+        {
+            Assert.NotNull(type.GetProperty(name));
```
Also. this is testing GerProperty with `BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public`. Is it worth testing `GetProperty` with e.g. `BindingFlags.Public | BindingFlags.NonPublic`?

jkotas replied:
Same for `GetProperties`  - test `BindingFlags.Public`, `BindingFlags.Public | BindingFlags.NonPublic`, `BindingFlags.NonPublic`?

AaronRobinsonMSFT committed:
Rename IsInterface to avoid Reflection issues

This is a narrow change to address a user
reported issue on System.Runtime and the
common implementation detail, System.RuntimeType.

AaronRobinsonMSFT commented:
I'm closing this issue for now. It is going to require additional discussion for fixes to the underlying issue with Reflection.
